File: leetcode/JavaScript实现/1-99题/20-39题/0022_括号生成.js
// 1、回溯
/*
1、某一次递归终止时需要将当前字符存入数组
2、 字符任取一个位置左侧必 左括号>=右括号
3、每次递归除了需要传当前字符还需要记情当前左右括号数
*/
var generateParenthesis = function (n) {
    let ans = [];

    let backtrack = (cur, left, right) => {
        if (cur.length === 2 * n) {
            ans.push(cur);
            return;
        }

        if (left < n)
            backtrack(cur + '(', left + 1, right);
        if (right < left)   
            backtrack(cur + ')', left, right + 1);
    };

    backtrack("", 0, 0);

    return ans;
};

File: leetcode/1-99题/0022_括号生成.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        int left = 0, right = 0;
        vector<string> ans;

        backtrack(ans, "", n, left, right);
        
        return ans;
    }
    void backtrack(vector<string> &res, string path, int n, int left, int right) {
        if (right > left || left > n || right > n)
            return;
        if (left == right && left == n) {
            res.push_back(path);
            return;
        }
        backtrack(res, path + '(', n, left + 1, right);
        backtrack(res, path + ')', n, left, right + 1);
    }
};

